[InstCombine] Treat umax as select(icmp eq x, 0), 1, x) in binop select fold.

davemgreen opened the pull request:
There is an existing instcombine for folding:
`(A ? B : C) binop (A ? E : F) -> A ? (B binop E) : (C binop F)`.
However this will not combine if the select `(x>=1 ? x : 1)` has
been converted to a `umax(x, 1)`. This adds code to treat the umax
as a select, allowing binops to fold into the select.

This patch handles the more special case where the predicate is the limit
extreme (such as `x==0 ? x : 1`), meaning it will use a ==0 predicate as
opposed to the >= predicate usually applied to a umax. According to alive this
is valid so long as the variables are not undef, are known to be non-equal
(https://alive2.llvm.org/ce/z/VAYw3J) or known to be the same value
(https://alive2.llvm.org/ce/z/4TSd7v).

Here are some proofs for different min/maxes:
umax: https://alive2.llvm.org/ce/z/kLw4J9
umin: https://alive2.llvm.org/ce/z/nvPa7p
smin: https://alive2.llvm.org/ce/z/HS8V7V
smax: https://alive2.llvm.org/ce/z/D2ocju

davemgreen committed:
[Instcombine] Add tests for treating umax as a select. NFC

nikic reviewed (commented):
Why is this specific to umin(1)? Doesn't the same work for, say, umin(2)? https://alive2.llvm.org/ce/z/gQqtpt

Or really any min/max intrinsic decomposed into select form.

davemgreen commented:
Hi - thanks for taking a look and sorry this took a while to get back to. This was causing some regressions in my motivating case due to knock-on register allocation. That should hopefully be resolved now and so it leads to an improvement as you would expect.

I've been looking at trying to generalize it to more cases and deal with the undef. I'll update this patch with the more special case, which the general case can be added to but is more complex. I think I have it working, but have to check through some of the details.

Unfortunately it appears some cases are not valid with undef arguments. I've tried to limit it to valid cases (as opposed to adding freeze).  I know that undef can be odd at times but I'm not sure it should be this odd.

davemgreen committed:
[InstCombine] Treat umax as select(icmp eq x, 0), 1, x) in binop select fold.

There is an existing instcombine for folding:
`(A ? B : C) binop (A ? E : F) -> A ? (B binop E) : (C binop F)`.
However this will not combine if the select `(x>=1 ? x : 1)` has
been converted to a `umax(x, 1)`. This adds code to treat the umax
as a select, allowing binops to fold into the select.

This patch handles the more special case where the predicate is the limit
extreme (such as `x==0 ? x : 1`), meaning it will use a ==0 predicate as
opposed to the >= predicate usually applied to a umax. According to alive this
is valid so long as the variables are not undef, are known to be non-equal
(https://alive2.llvm.org/ce/z/VAYw3J) or known to be the same value
(https://alive2.llvm.org/ce/z/4TSd7v).

Here are some proofs for different min/maxes:
umax: https://alive2.llvm.org/ce/z/kLw4J9
umin: https://alive2.llvm.org/ce/z/nvPa7p
smin: https://alive2.llvm.org/ce/z/HS8V7V
smax: https://alive2.llvm.org/ce/z/D2ocju

davemgreen commented:
The more general cases will hopefully be something like https://github.com/davemgreen/llvm-project/commit/7ba90fb79a0df6c9254079ebb993ace400587f58.

nikic commented:
I'm not a fan of the special undef handling here. We should either insert the necessary freeze, or, if that covers your motivating case, we can start with guarding the transform under isGuaranteedNotToBeUndefOrPoison().
